### DISCUSSION TITLE
Building the Alpha NT cross-compiler & libraries

### DISCUSSION THREAD

**Discussion by davidrg** [open-watcom/open-watcom-v2#946] [Q&A]
Can anyone provide any hints on how to build a working cross-compiler for Alpha? I managed to do it [years ago with OpenWatcom 1.7](http://www.ext.zx.net.nz/software/notes/watcom-axp/) but things seem to have changed since then. I can produce claxp, linkaxp and asaxp by uncommenting things in \bld\mstools\builder.ctl but I'm not sure how to produce the libaxp directory. Last time I could just run `pmake -d cpu_axp -h` followed by `builder rel2 cpu_axp` in each of {clib, mathlib, w32api, dwarf, cg, cfloat, as, owl, plusplus} but pmake seems to be gone now.

I ran a build of my current project with the wccaxp that's included in current builds and to my surprise it actually got all the way through to the link step before falling over so I'm currently wondering if I might actually get a working Alpha NT binary out if I could figure out how to build the libraries and other bits.

**Comment by jmalak** [comment]
I don't think it is necessary to manipulate with builder control files, all should be there, if not then raise issue.
For AXP architecture and Windows NT you should use following steps:
- in OW root folder to run command **`build preboot`**, it should build wmake and builder tools
- in OW **`bld`** folder to run command **`builder boot`**, it should create OW bootstrap tools for your host
- in OW **`bld`** folder to run command **`builder rel cpu_axp os_nt .and .or`**, it should create OW build including AXP stuff

**Reply by davidrg** [reply]
I've had a go using the code from the 2022-08-1 build and made a little progress. For some reason linkaxp seems to insist on looking for libraries in lib386 rather than libaxp - I didn't see any options to tell it to use the AXP libaries so I just renamed libaxp to lib386 to see if my app was going to link at all, which it didn't. Though my collection of link errors _has_ been reduced to just a handful now - things like __imp_fprintf, __imp_exit, __imp_fgetc, etc. Not sure if there is still some AXP library I'm missing, if there is some stuff missing from the AXP libraries, or if its just this code doing something weird and redefining things (its very old and it supports a lot of platforms, though it does built for x86 Windows and OS/2 with OpenWatcom fine). 

Getting the AXP bits built was quite involved - some stuff is commented out in builder.ctl files and other stuff is missing entirely or just doesn't build at all so has to be disabled. This is what I ended up doing to get what *looks* like a complete libaxp directory:

1. run: `builder preboot`
2. in \bld, run: `builder boot`
3. in \bld, run: `builder rel` - this is required to get wccaxp.exe and wppaxp.exe
4. copy `\rel\binnt\wccaxp.exe` to `\build\binbuild\bwccaxp.exe`
   copy `\rel\binnt\wppaxp.exe` to `\build\binbuild\bwppaxp.exe`
   copy `\rel\binnt\{wccaxp.dll,wppdaxp.dll}` to `\build\binbuild\`
5. Edit `bld\builder.ctl` and comment out:
			line 148: `[ INCLUDE "<OWROOT>/bld/f77/f77lib/builder.ctl" ]`
			line 176: `[ INCLUDE "<OWROOT>/bld/f77/wfc/builder.ctl" ]`
   Because building `f77\wfc\axp\ntaxp` fails with: `Error(F38): (wasmcmsg.grh) does not exist and cannot be made from existing files`
6. Edit `bld\mstools\builder.ctl` and uncomment:
			line 18: `#    <CCCMD> asaxp/generic/nt386/<OWOBJDIR>/asaxp.exe    "<OWRELROOT>/binnt/"`
			line 21: `#    <CCCMD> cl/axp/nt386/<OWOBJDIR>/claxp.exe           "<OWRELROOT>/binnt/"`
			line 25: `#    <CCCMD> lib/axp/nt386/<OWOBJDIR>/libaxp.exe         "<OWRELROOT>/binnt/"`
			line 28: `#    <CCCMD> link/axp/nt386/<OWOBJDIR>/linkaxp.exe       "<OWRELROOT>/binnt/"`
7. Edit `bld\w32api\builder.ctl` and uncomment:
			line 29: `#    <CCCMD> ntaxp/lib/*                   "<OWRELROOT>/libaxp/nt/"`
8. Rename the following files to something else to prevent the build failing 
    with Error opening file ../../master.mif - no such file or directory			
			`\bld\as\axp\ntaxp\test\makefile`
			`\bld\as\ppc\ntaxp\test\makefile`
9. Edit `\bld\cc\builder.ctl` and comment out the following:
			line 13: `[ INCLUDE mps/builder.ctl ]`
			line 14: `[ INCLUDE ppc/builder.ctl ]`
	These don't build - for mps/builder.ctl, you get:
			`Error(F38): (..\..\..\..\bld\as\mps\ntaxp.dll\asinline.lib) does not exist and cannot be made from existing files`
	And for `ppc/builder.ctl`, you get:
			`Error(F38): (..\..\..\..\bld\cg\risc\ppc\ntaxp.dll\cgppc.lib) does not exist and cannot be made from existing files`
101. In \bld, run: `builder rel cpu_axp os_nt .and .or`

Eventually it failed in \bld\dip\dwarf\ntaxp with: ..\c\stub.c(59): Warning! W131: No prototype found for function 'DllMainCRTStartup'

I've not bothered trying to get it any further at this point as it looks like the rest of the stuff its going to try and build isn't important for cross-compiling from windows. From here I had to fetch a handful of libs out of \bld myself as I guess they're missing from a builder.ctl somewhere: mthdll.lib, clbdll.lib, plbdll.lib, plbdllx.lib and wrtdll.lib (and the associated dll files)

So overall it looks promising. The compiler at least _seems_ to work fine - no crashes due to missing features in the code generator like in OpenWatcom 1.7.

**Comment by jmalak** [comment]
Sorry, you don't get here info what host system and architecture you use for build.
Because it is not clear if you are cross compiling from another architecture or use native build on AXP.
If you are cross-compiling then you use incorrect procedure.
If you are on native AXP system then it is bug.

**Reply by davidrg** [reply]
I was hoping to make a cross-compiler that can target Alpha NT from an x86 windows box. I was using OpenWatcom 1.9 to compile OpenWatcom 2.0

I did initially try following the three steps you supplied but the third step fails pretty quickly complaining that bwccaxp.exe can't be found - it seems neither of the previous steps generates it. If I give it a wccaxp.exe from a full build it will later run into the same problem with bwppaxp.exe.

With those two files in place it will continue building for a little while before failing due to problems with some of the libraries and other bits. Some things are just never copied to the output without modifying builder.ctl files as they're commented out - eg, [claxp.exe](https://github.com/open-watcom/open-watcom-v2/blob/master/bld/mstools/builder.ctl#L21) and the [win32 libraries](https://github.com/open-watcom/open-watcom-v2/blob/master/bld/w32api/builder.ctl#L29) are commented out for Alpha

**Comment by jmalak** [comment]
Sorry, I didn't do such non-standard build very long time.
I tried it just now and build system will require to do some fixes or changes.
I submit change to include bootstrap cross-compilers  to standard build that no problem with compilers for AXP/PPC/MIPS architectures.
But more fixes will be necessary.

**Comment by kristibektashi** [comment]
Now that Windows NT 4 on Alpha can be emulated (https://betawiki.net/wiki/Draft:Installing_Windows_NT_on_DEC_Alpha), this may be due for an update

**Reply by kristibektashi** [reply]
Ping @jmalak @davidrg
